File: 2020/02/main.py
import os
import re
import sys

from lib.executor import Executor

CWD = os.path.dirname(os.path.abspath(__file__))


def valid_part1(min, max, char, pwd):
    num_char = 0
    for c in pwd:
        if c == char:
            num_char += 1
            if num_char > max:
                return False
    return num_char >= min


def valid_part2(pos1, pos2, char, pwd):
    return (pwd[pos1 - 1] == char) != (pwd[pos2 - 1] == char)


def num_valid(input, valid_fn):
    num_correct = 0
    for line in input:
        match = re.match(r"^(\d+)-(\d+) (\w): (\w+)$", line)
        if valid_fn(int(match[1]), int(match[2]), match[3], match[4]):
            num_correct += 1
    return num_correct


def part1_solution(input):
    return num_valid(input, valid_part1)


def part2_solution(input):
    return num_valid(input, valid_part2)


executor = Executor(
    [l.rstrip() for l in open(f"{CWD}/input.txt", "r").readlines()],
    part1_solution,
    part2_solution,
)
executor(sys.argv)

File: 2020/02/main.rb
require_relative '../../lib/ruby/executor'

module Part1Validator
  def self.valid?(min, max, char, pwd)
    num_chars = pwd.chars.count { |c| c == char }
    num_chars >= min && num_chars <= max
  end
end

module Part2Validator
  def self.valid?(pos1, pos2, char, pwd)
    (pwd[pos1 - 1] == char) != (pwd[pos2 - 1] == char)
  end
end

def count_valid(input, validator)
  input.count do |line|
    match = line.match(/^(\d+)-(\d+) (\w): (\w+)$/)
    validator.valid?(Integer(match[1]), Integer(match[2]), match[3], match[4])
  end
end

part1_proc = Proc.new { |input| count_valid(input, Part1Validator) }
part2_proc = Proc.new { |input| count_valid(input, Part2Validator) }
executor = Executor.new(
  File.readlines(File.join(File.dirname(__FILE__), 'input.txt')).map(&:rstrip),
  part1_proc,
  part2_proc
)
executor.run(ARGV)
